### PR TITLE
[Fix #1453] Re-enable auto-save-mode

### LINKED ISSUE #1453: Add auto-save feature for Spacemacs
Here is [a nifty trick for auto saving when switching window or buffer](http://batsov.com/articles/2012/03/08/emacs-tip-number-5-save-buffers-automatically-on-buffer-or-window-switch/).

Then, we can turn on `auto-save-mode` and automatically save the current active buffer with this function add to `auto-save-hook`:

```elisp
   (defun save-buffer-if-visiting-file (&optional args)
      "Save the current buffer only if it is visiting a file"
      (interactive)
      (if (and (buffer-file-name) (buffer-modified-p))
          (save-buffer args)))

   (add-hook 'auto-save-hook 'save-buffer-if-visiting-file)
```

We need the hook because despite the name `auto-save-mode`, it does something else: it automatically saves your buffer, but to a backup file rather than directly on your editing buffer.

### PR DESCRIPTION
**Pull request by tuhdo**
- Re-enable
- Move all autosave files to spacemacs-cache-directory
- Allow Auto-saving directly on current buffer.

### CODE REVIEW

**Comment by ralesi** [comment]
Can you make the auto save hook toggled by a dotspacemacs variable?  I dont like to save unless I explicitly ask to since persistent-undo is disabled

**Comment by syl20bnr** [comment]
@tuhdo can you explain what this feature does exactly ? Does it create a temporary file in .cache or does it save the actual file ?

**Comment by tuhdo** [comment]
with this PR, it does both. Without the hook, it only save to the cache.

**Comment by tuhdo** [comment]
@syl20bnr I updated it.

**Comment by rishihahs** [comment]
+1 Merging this PR would be great.

**Comment by ralesi** [comment]
+1 on merging as well

**Comment by tuhdo** [comment]
@syl20bnr we should re-enable `auto-save-mode` to avoid users losing data.

File: core/core-dotspacemacs.el [hunk 1]
```diff
@@ -83,6 +83,10 @@ size to make separators look not too crappy.")
 By default the command key is `:' so ex-commands are executed like in Vim
 with `:' and Emacs commands are executed with `<leader> :'.")
 
+(defvar dotspacemacs-autosave-file-directly nil
+  "If non-nil, auto-save to the file directly. If nil, just save
+to auto-save file in `spacemacs-cache-directory'")
+
```
Comment (syl20bnr) [inline core/core-dotspacemacs.el]: Sorry for the delay to review this.

I did not understand the meaning of this variable, I only understood it by looking at where it is used in the code. We have te rename it.

`dotspacemacs-auto-save-new-buffer-to-file`

**Comment by tuhdo** [comment]
It does not save new buffer to file, but save the current buffer directly to file of that buffer, hence the name. Think this is like your regular editor with auto-save on, then the file is saved directly. Auto-save in Vim/Emacs works different: it saves to some other file (i.e. like Vim swap file), not the current file.

**Comment by syl20bnr** [comment]
It need to be more explicit, using the negation is maybe more explicit in this case:

`dotspacemacs-auto-save-to-temp-file` which default to `t`.

**Comment by syl20bnr** [comment]
the `directly` is what sounds wrong, it is not enough specific. It convey a notion of time (which made me think if was for new buffers, even after looking at the code) whereas you want a notion of location/name.

**Comment by syl20bnr** [comment]
`dotspacemacs-auto-save-to-different-file`

**Comment by syl20bnr** [comment]
We may improve this dotfile variable by allowing several values:

`dotspacemacs-auto-save-file-to`:
- `'same-file`
- `'temp-file`
- `nil` = no auto-save

**Comment by syl20bnr** [comment]
Thank you ! :+1: 

I made the implementation mentioned above but slightly differently.
The variable is `dotspacemacs-auto-save-file-location` and it takes the followin values:
- `original` auto-save to the original file
- `cache` auto-save in cache directory `.cache/auto-save/site/`
- `nil` disable auto-save

For the `original` part we don't need the hook, setting the variable `auto-save-visited-file-name` to `t` does it.

I removed some code about auto-saving that were put at another place in the `config.el` file.

I put back the entry to auto-save remote URLs on the local machine, I set the directory to `.cache/auto-save/dist/`

More info in related commit: https://github.com/syl20bnr/spacemacs/commit/cd70f04

Cherry-picked into `develop`, you can safely delete your branch.

**Comment by dijonkitchen** [comment]
While this should auto-save based on time, I don't think it saves immediately on context change (switching to a new buffer or window).

**Pull request closed without merging**